common-money.lic

```diff
@@ -21,11 +40,29 @@ module DRCM
   end
 
   def convert_to_copper(amount, denomination)
-    return (amount.to_i * 10_000) if 'platinum' =~ /^#{denomination}/
-    return (amount.to_i * 1000) if 'gold' =~ /^#{denomination}/
-    return (amount.to_i * 100) if 'silver' =~ /^#{denomination}/
-    return (amount.to_i * 10) if 'bronze' =~ /^#{denomination}/
-    amount.to_i
+    # Convert to copper given denomination (abbreviation permitted)
+    # If no denomination specified, return the integer amount (assumed to be coppers)
+    denomination = denomination.strip # trim whitespace and also convert nil to empty string
+    if !denomination.empty?
```

Comment:
This is a change in behavior in the case where denomination is not-empty and also fails to match any of the checks. E.g. someone passes in `glod` or `sliver`. Do you think that case should also return `amount.to_i`?

Reply:
This is a good question. 

If I'm not mistaken, the original accepted a denomination arg. If it matched the full DR denominations, it did conversions. If it didn't match (coper, sliver, etc.), or was nil, it integerized the amount and returned it.

I believe my first revision unintentionally changed that behavior. If denomination was not nil, it checked for matches, including substrings, to the DR denominations. If denomination WAS nil, it integerized the amount arg, and returned it. But if the denomination was existed but was misspelled, for example, I'm betting my revision would do nothing at all.

Now this revision per @KatoakDR  is a change in behavior from MY revision, but I think behaves similarly to the original, with the added advantage that it now warns the user that the passed-in denomination arg is not known, and still completes the function by integerizing the amount arg and returning it.

I kind of like Katoak's revision best. But now that we've had some discussion on it, I could easily mirror the original's behavior if that's best.

Reply:
> This is a change in behavior in the case where denomination is not-empty and also fails to match any of the checks

Good catch. To restore that original behavior then I think we just need to move the default logic out of the ELSE so it always runs if the script gets to that point. In the below proposal, if the denomination is not blank and does not match the platinum/gold/silver/bronze/copper checks then it hits the default case and will message the warning and assume coppers.

```ruby
  def convert_to_copper(amount, denomination)
    denomination = denomination.strip # trim whitespace and also convert nil to empty string
    if !denomination.empty?
      return (amount.to_i * 10_000) if 'platinum'.start_with?(denomination.downcase)
      return (amount.to_i *   1000) if 'gold'.start_with?(denomination.downcase)
      return (amount.to_i *    100) if 'silver'.start_with?(denomination.downcase)
      return (amount.to_i *     10) if 'bronze'.start_with?(denomination.downcase)
      return (amount.to_i *      1) if 'copper'.start_with?(denomination.downcase)
    end
    DRC.message("Unknown denomination, assuming coppers: #{denomination}")
    amount.to_i
end
```

Reply:
Returned to original behavior.  Test:

```
>;tip gil 10 floof
--- Lich: tip active.
[tip: Denomination is floof]
[tip: Currency is ]
| Unknown denomination, assuming coppers: floof
[tip]>wealth

Wealth:
  3 silver Kronars (300 copper Kronars).
  No Lirums.
  2 gold, and 39 copper Dokoras (2,039 copper Dokoras).
>
[tip]>tip gil 10 dokoras
Hmmm, tipping yourself?  That's an interesting concept, but it won't work.```